### PR TITLE
factory reset

### PR DESCRIPTION
**Pull request by JarbasAl**
## events
- trigger with event `system.factory.reset`
- individual PHAL plugins register with `system.factory.reset.register`
   - on `__init__` method
   - on `system.factory.reset.ping`
- `system.factory.reset.start` emitted
- `system.factory.reset.phal` emitted
- wait for all registered plugins to emit `system.factory.reset.phal.complete` or timeout after 60 seconds
- emit `system.factory.reset.complete`
- emit `system.reboot`

### CODE REVIEW

**Comment by AIIX** [comment]
as discussed in chat if we want shell to handle parsing the script we need to add two more settings flags apart from script path, external=true / false and an external bus message to the app (shell here with the script path) if external app handling is true

**Comment by JarbasAl** [comment]
example plugin integration https://github.com/OpenVoiceOS/ovos-PHAL-plugin-mk1/commit/fa5b5242fa1b3e6fd6cec08ce53b7ffd55a27f50

**Comment by j1nx** [comment]
Nit-picking here but this is not a real factory reset but more an ovos-core reset.

In my mind a factory reset means, going back to the exact same machine state as freshly flashed.

**Comment by JarbasAl** [comment]
@j1nx that is plarform dependent, and thats exactly what the bash script is there for, a distro can specify an extra bash script to do platform specific resets, the plugin just ensures the bare minimum is done, eg, for a desktop install

other plugins can also do reset actions, see the mk1 example to reset faceplate hardware

**Comment by j1nx** [comment]
Then I thingk I missread the whole code. I thought it is also taking care of deleting things and rebooting. Did not see (still not see) the bash stuff but it is OK. My python still sucks..

For the ovos-buildroot image, a factory reset just means to reboot with a cmdine flag which at pre-init nukes the whole overlayfs system and reboots again. Then everything is back at flash time.

Will reach out later how to implement that later.

File: ovos_PHAL_plugin_system/__init__.py [hunk 1]
```diff
@@ -25,11 +34,107 @@ def __init__(self, bus=None, config=None):
...
+            self.bus.on("system.factory.reset.phal.complete", on_done)
+            self.bus.emit(message.forward("system.factory.reset.phal"))
+            event.wait(timeout=60)
+            self.bus.remove("system.factory.reset.phal.complete", on_done)
+
+        script = message.data.get("script", True)
+        if script:
+            script = os.path.expanduser(self.config.get("reset_script", ""))
+            if os.path.isfile(script):
+                subprocess.call(script, shell=True)
```
Comment (JarbasAl) [inline ovos_PHAL_plugin_system/__init__.py:130]: @j1nx this is where any platform specific code would be executed

Comment (JarbasAl) [reply ovos_PHAL_plugin_system/__init__.py:130]: so under PHAL config under this plugin section just need to add `"reset_script": "path/to/custom/code.sh"`

Comment (j1nx) [reply ovos_PHAL_plugin_system/__init__.py:130]: And then it only just rans what is in there?

Comment (JarbasAl) [reply ovos_PHAL_plugin_system/__init__.py:130]: Yes

Comment (JarbasAl) [reply ovos_PHAL_plugin_system/__init__.py:130]: Not only, the bus message can request specific actions only

**Pull request merged**